src/v/config/property.h

```diff
@@ -625,8 +629,9 @@ class retention_duration_property final
     // serialize the value. the key is taken from the property name at the
     // serialization point in config_store::to_json to avoid users from being
     // forced to consume the property as a json object.
-    void to_json(
-      json::Writer<json::StringBuffer>& w, redact_secrets) const final {
+    void
+    to_json(json::Writer<json::StringBuffer>& w, redact_secrets) const final {
+        vassert(!is_secret(), "{} must not be a secret", name());
```

Comment:
maybe a comment here would be useful. for example, it seems as though the property itself can be secret or non-secret (see print() above), but what keeps this method from being invoked on an instance of the property that is secret? it seems odd that `retention_duration_property` is special in a way that it must not be a secret, but there aren't others instances of requirement. for example, why is this assertion made

Reply:
That's a good point. I suspect `print()` is defined as such because it might've copied from the vanilla `property::print()` method. I made this change at John's request to avoid serializing the wrong type to the returned JSON. That said, you're right that there isn't anything that forces us to ensure this is always non-secret.

A few options:
- stick the `vassert` in a new constructor for `retention_duration_property` that just passed through to vanilla `property`'s constructor
- revert this and serialize potentially surprising JSON
- continue as is but comment that we're just trying to avoid returning surprising JSON

I'm leaning towards the first option. Would that be less surprising code (with a comment)?

Reply:
Actually went with the last option, and updated `print()` accordingly